src/script/script_exec_data.h

```diff
@@ -5,19 +5,48 @@
 #ifndef BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
 #define BITCOIN_SCRIPT_SCRIPT_EXEC_DATA_H
 
+#include <hash.h>
+#include <script/script.h>
 #include <uint256.h>
 
 /**
  * Struct for holding data generated during script execution which is used
  * during script execution, e.g. for computing sig hashes.
  */
 struct ScriptExecutionData {
+    static constexpr uint32_t DEFAULT_CODESEP_POS = 0xffff'ffff;
+
     /**
      * Opcode position of the last executed OP_CODESEPARATOR.
-     * 
+     *
      * This allows signatures to commit to certain code paths.
      */
     uint32_t m_codeseparator_pos;
+
+    /**
+     * SHA256 of the (complete) script being executed, regardless of
+     * OP_CODESEPARATOR.
+     *
+     * - In plain scripts, this is the SHA256 of the scriptPubKey.
+     * - In P2SH, this is the SHA256 of the redeemScript.
+     * - In Tapscripts, this is the "tapleaf hash", calculated as defined in
+     *   BIP341: taggedhash_TapLeaf(version || compact_size(len(script)) ||
+     *                              script).
+     * - In Taproot, this entire struct would not be constructed.
+     */
+    std::optional<uint256> m_executed_script_hash;
+
+    /**
+     * ScriptExecutionData with the SHA-256 of executed_script and as
+     * codeseparator position.
+     */
+    static ScriptExecutionData
+    SHA256Script(const CScript &executed_script,
```

Comment:
Does this relate to `m_executed_script_hash` above?

Reply:
I think we should define a non-default constructor here, and use this in the constructor.  Require a `CScript`, but it would just be empty for the initial case of executing the scriptSig.

Reply:
This will prevent the inadvertant use here without defining the m_executed_script_hash.